File: src/python/core/application/services/primary_skills_extractor.py
from typing import List, Optional
from ...domain.entities.candidate_record import CandidateRecord
from ...domain.enums.skill_level import SkillLevel
from ...domain.configuration.vector_metadata_config import VectorMetadataConfig


class PrimarySkillsExtractor:
    def __init__(self, config: VectorMetadataConfig):
        self.config = config
    
    def extract(self, candidate: CandidateRecord) -> List[str]:
        if not candidate.SkillMatrix:
            return []
        
        primary_skills = []
        
        for skill in candidate.SkillMatrix:
            if not skill.SkillName:
                continue
            
            if skill.SkillLevel in self.config.strong_skill_levels:
                primary_skills.append(skill.SkillName)
        
        return primary_skills[:self.config.primary_skills_max_count]

File: src/dotnet/Core/Application/Services/PrimarySkillsExtractor.cs
using Rag.Candidates.Core.Domain.Configuration;
using Rag.Candidates.Core.Domain.Entities;

namespace Rag.Candidates.Core.Application.Services;

public sealed class PrimarySkillsExtractor
{
    private readonly VectorMetadataConfig _config;

    public PrimarySkillsExtractor(VectorMetadataConfig config)
    {
        _config = config;
    }

    public string[] Extract(CandidateRecord candidate)
    {
        if (candidate.SkillMatrix == null || candidate.SkillMatrix.Length == 0)
            return Array.Empty<string>();

        var primarySkills = new List<string>();

        foreach (var skill in candidate.SkillMatrix)
        {
            if (string.IsNullOrWhiteSpace(skill.Name))
                continue;

            // Check if skill level is in the strong category
            if (skill.Level != null && _config.StrongSkillLevels.Contains(skill.Level))
            {
                primarySkills.Add(skill.Name);
            }
        }

        // Limit to max count and return
        return primarySkills.Take(_config.PrimarySkillsMaxCount).ToArray();
    }
}
